File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.io.*;
import java.util.*;

public class Main {
    static boolean[] visit;
    static int[] num;
    static StringBuilder sb=new StringBuilder();
    public static void main(String[] args) throws IOException {
        BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st=new StringTokenizer(br.readLine());
        int N=Integer.parseInt(st.nextToken());
        int M=Integer.parseInt(st.nextToken());

        num=new int[M];
        visit=new boolean[N];

        dfs(N, M, 0);
        System.out.println(sb);

        br.close();
    }
    static void dfs(int N, int M, int depth){
        if (depth==M){
            for (int val: num){
                sb.append(val).append(' ');
            }
            sb.append('\n');
            return;
        }

        for (int i=0; i<N; i++){
            if (!visit[i]){
                visit[i]=true;
                num[depth]=i+1;
                dfs(N,M,depth+1);
                visit[i]=false;
            }
        }
        return;
    }
}

File: C++17/백준/Silver/15649. N과 M （1）/N과 M （1）.cc
#include <iostream>
using namespace std;

int N, M;
bool visited[9];
int num[9];

void dfs(int depth);

int main(){
    cin >> N >> M;
    dfs(0);
    return 0;
}

void dfs(int depth){
    if (depth == M){
        for (int i = 0; i < M; i++){
            cout << num[i] << ' ';
        }
        cout << '\n';
        return;
    }

    for(int i = 0; i < N; i++){
        if(!visited[i]){
            visited[i] = 1;
            num[depth] = i + 1;
            dfs(depth + 1);
            visited[i] = 0;
        }
    }
}